### PR TITLE
Fix for incomplete json responses

### PR DESCRIPTION
**Pull request by slitviachenko**
## Description
Fix for incomplete json responses.

I created the `parseResponse` function that attempts to parse a JSON string, handling cases where the JSON might be incomplete. If the initial parsing fails, it tries to fix the JSON string by checking its structure (`array` or `object`) and attempting to complete it. If it still can't parse the string after attempting to fix it, it returns the original incomplete string.

## Related Issue
https://jira.corp.adobe.com/browse/SITES-21991

## How Has This Been Tested?
You can use the following URL, I set `max_tokens: 200,` instead of the current `max_tokens: 800,` value to reproduce incomplete JSON responses.
https://experience-qa.adobe.com/?devMode=true&shell_source=workspace&workspace=sites21991brokenjson&shell_ims=prod#/aem/generate-variations/

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: web-src/src/helpers/ResultsParser.js [hunk 1]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+ *    - If no closing brace is found, returns an empty object `{}` for objects or an empty array `[]` for arrays.
+ */
+const parseResponse = (str) => {
+  const fixJson = (incomplete) => {
+    const firstChar = incomplete.substring(0, 1);
+    const isArray = firstChar === '[';
+    const isObject = firstChar === '{';
+    if (!isArray && !isObject) {
+      return incomplete;
+    } else {
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:40]: i think you can avoid using else here to improve code readability

File: web-src/src/helpers/ResultsParser.js [hunk 2]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+ * The function works as follows:
+ * 1. Tries to parse the input string using `JSON.parse`.
+ * 2. If parsing fails, it invokes an internal `fixJson` function:
+ *    - Checks the first character to determine if the JSON represents an array or object.
+ *    - Finds the last closing curly brace (`}`) in the string.
+ *    - Truncates the string up to the last closing brace and attempts to close it correctly.
+ *    - Parses the truncated string. If this fails, returns the original incomplete string.
+ *    - If no closing brace is found, returns an empty object `{}` for objects or an empty array `[]` for arrays.
+ */
+const parseResponse = (str) => {
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:33]: can you cover this function with unit test to validate that it works as expected for different inputs?

File: web-src/src/helpers/ResultsParser.js [hunk 3]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+    const firstChar = incomplete.substring(0, 1);
+    const isArray = firstChar === '[';
+    const isObject = firstChar === '{';
+    if (!isArray && !isObject) {
+      return incomplete;
+    } else {
+      const lastClosingCurlyBrace = incomplete.lastIndexOf('}');
+      if (lastClosingCurlyBrace === -1) {
+        return isArray ? [] : {};
+      } else {
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:44]: as above, you can avoid using else

File: web-src/src/helpers/ResultsParser.js [hunk 4]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+ */
+const parseResponse = (str) => {
+  const fixJson = (incomplete) => {
+    const firstChar = incomplete.substring(0, 1);
+    const isArray = firstChar === '[';
+    const isObject = firstChar === '{';
+    if (!isArray && !isObject) {
+      return incomplete;
+    } else {
+      const lastClosingCurlyBrace = incomplete.lastIndexOf('}');
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:41]: i think you can modify this condition to satisfy array and obj use cases:
```suggestion
     const lastClosingCurlyBrace =  isArray ? incomplete.lastIndexOf(']') : incomplete.lastIndexOf('}')
```

File: web-src/src/helpers/ResultsParser.js [hunk 5]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+      if (lastClosingCurlyBrace === -1) {
+        return isArray ? [] : {};
+      } else {
+        let truncated = incomplete.substring(0, lastClosingCurlyBrace + 1);
+        truncated += isArray ? ']' : '}';
+        try {
+          return JSON.parse(truncated);
+        } catch (e) {
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:49]: let's log this exception so we can track what use cases cause errors

File: web-src/src/helpers/ResultsParser.js [hunk 6]
```diff
@@ -14,9 +14,55 @@ function objectToString(obj) {
...
+          return incomplete;
+        }
+      }
+    }
+  };
+
+  try {
+    return JSON.parse(str);
+  } catch (err) {
+    return fixJson(str);
```
Comment (irenelagno) [inline web-src/src/helpers/ResultsParser.js:59]: // If parsing fails, attempt to fix and parse again
add here logging as well

**Review by irenelagno — CHANGES_REQUESTED** [review]

**Comment by irenelagno** [comment]
This response from the LLM is misformatted. I can got it when I click through UI on create new prompt
Screen Shot 2024-06-03 at 11 14 21 AM
So, it seems we need to add some check for this use case. 
`{
  "jobId": "20e8b93da1d04e8aa8b93da1d09e8a73",
  "result": {
    "context": null,
    "conversation_identifier": 609261,
    "generations": [[{
      "generation_info": {
        "content_filter_results": {
          "hate": {
            "filtered": false,
            "severity": "safe"
          },
          "self_harm": {
            "filtered": false,
            "severity": "safe"
          },
          "sexual": {
            "filtered": false,
            "severity": "safe"
          },
          "violence": {
            "filtered": false,
            "severity": "safe"
          }
        },
        "finish_reason": "stop"
      },
      "message": {
        "additional_kwargs": {

        },
        "content": "The cinema has a 30% discount on Mondays. Therefore, John needs to pay only 70% of the ticket price. \n\nSo, to calculate the actual price John paid, you first find 70% of the $28 ticket price:\n\n$28 * 0.70 = $19.60",
        "example": false,
        "type": "ai"
      },
      "text": "The cinema has a 30% discount on Mondays. Therefore, John needs to pay only 70% of the ticket price. \n\nSo, to calculate the actual price John paid, you first find 70% of the $28 ticket price:\n\n$28 * 0.70 = $19.60",
      "type": "ChatGeneration"
    }]],
    "llm_output": {
      "model_name": "gpt-4",
      "prompt_filter_results": [{
        "content_filter_results": {

        },
        "prompt_index": 0
      }],
      "token_usage": {
        "completion_tokens": 62,
        "prompt_tokens": 7,
        "total_tokens": 69
      }
    },
    "llm_type": "azure_chat_openai",
    "messages": null,
    "query_id": 604226
  },
  "status": "completed"
}`

File: web-src/src/helpers/ResultsParser.js (hunk 4 above)
Comment (slitviachenko) [reply web-src/src/helpers/ResultsParser.js:41]: Right now, we should build prompts in a way that ensures we get a response as an array of objects. Therefore, an additional check for whether we get an object might be unnecessary. However, it doesn't hurt to have it.

**Comment by slitviachenko** [comment]
@irenelagno , I've addressed all your comments and redeployed the changes. Could you please take one more look whenever you have a chance?

**Review by vtsaplin — APPROVED** [review]

**Review by irenelagno — APPROVED** [review]
LGTM

**Pull request merged**